File: 2131. Longest Palindrome by Concatenating Two Letter Words.py
from collections import Counter
class Solution:
    def longestPalindrome(self, words: List[str]) -> int:
        counter = Counter(words)
        center = 0
        palin_diff = 0
        palin_same = 0
        
        for word, count in counter.items():
            if word[0]==word[1]:
                palin_same += count//2*2
                
                if count%2==1: center =2
            
            else:
                palin_diff += min(counter[word], counter[word[::-1]])*0.5
        
        return palin_same*2 + int(palin_diff)*4 + center
                
                
                

File: 2131. Longest Palindrome by Concatenating Two Letter Words.java
class Solution {
    public int longestPalindrome(String[] words) {
        HashMap<String, Integer> m = new HashMap();
        
        for (String word: words) {
            if (!m.containsKey(word)) m.put(word,0);
            m.put(word,m.get(word)+1);
        }
        
        int palin_diff = 0;
        int palin_same = 0;
        int center = 0;
        
        for (String word: m.keySet()){
            if (word.charAt(0) == word.charAt(1)){
                palin_same = palin_same + m.get(word)/2*2;
                if (m.get(word)%2 == 1) {
                center = 2;
                }
            }
            else{
                String reverse = Character.toString(word.charAt(1)) + Character.toString(word.charAt(0));
                if (m.containsKey(reverse)){
                    palin_diff = palin_diff + Math.min(m.get(word), m.get(reverse));
                }
            }
        }

        return palin_same*2 + palin_diff*2 + center;
    }
}